### PR TITLE
pool: support creation by means of URIs

### PR DESCRIPTION
**Pull request by gila**
This change will create base bdevs using URIs if we are passed
a valid scheme. If we use the legacy model where we specify
a device path like /dev/sdx -- we will use that

### CODE REVIEW

File: mayastor/src/pool.rs [hunk 1]
```diff
@@ -419,42 +420,60 @@ impl Pool {
...
 
-        let (sender, receiver) = oneshot::channel::<i32>();
-        if base_bdev_type == "aio" {
-            unsafe {
-                bdev_aio_delete(
-                    base_bdev.as_ptr(),
-                    Some(done_cb),
-                    cb_arg(sender),
-                );
+            let (sender, receiver) = oneshot::channel::<i32>();
```
Comment (tiagolobocastro) [inline mayastor/src/pool.rs:442]: Do we need this "legacy" stuff or is this just a stop gap until moac creates pools with URIs?

File: mayastor/src/core/bdev.rs [hunk 2]
```diff
@@ -115,13 +115,28 @@ impl Share for Bdev {
...
-    fn get_share_uri(&self) -> Option<String> {
+    fn share_uri(&self) -> Option<String> {
         match self.shared() {
             Some(Protocol::Nvmf) => nvmf::get_uri(&self.name()),
             Some(Protocol::Iscsi) => iscsi::get_uri(Side::Nexus, &self.name()),
             None => None,
         }
     }
+
+    fn bdev_uri(&self) -> Option<String> {
```
Comment (tiagolobocastro) [inline mayastor/src/core/bdev.rs:126]: hmm I was indeed torn between adding it here or as a trait when I added the Bdev->Url converter.

File: mayastor/src/nexus_uri.rs [hunk 3]
```diff
@@ -83,6 +83,21 @@ pub fn bdev_get_name(uri: &str) -> Result<String, NexusBdevError> {
...
+                    }
+            }
+            _ => false,
+        }
+    }
+}
+
 impl std::cmp::PartialEq<url::Url> for Bdev {
     fn eq(&self, uri: &url::Url) -> bool {
         match Uri::parse(&uri.to_string()) {
```
Comment (tiagolobocastro) [inline mayastor/src/nexus_uri.rs:103]: since you've added the trait above, this can just become self == uri ?

File: mayastor/src/pool.rs [hunk 4]
```diff
@@ -560,6 +583,63 @@ pub(crate) async fn create_pool(
...
+        bdev_if: e.to_string(),
+        name: args.disks[0].clone(),
+    })?;
+
+    if let Some(pool) = Pool::lookup(&args.name) {
+        return if pool.get_base_bdev().name() == parsed.get_name() {
+            Ok(pool.into())
+        } else {
+            Err(Error::AlreadyExists {
+                name: args.name,
```
Comment (tiagolobocastro) [inline mayastor/src/pool.rs:607]: should this error include the existing disk as well?

File: mayastor/src/pool.rs (hunk 1 above)
Comment (gila) [reply mayastor/src/pool.rs:442]: Right. We will remove the "legacy" stuff altogether but later one, but for that the whole pool creation would need to change somewhat, beginning with examining the bdevs. I want to postpone that until later.

File: mayastor/src/nexus_uri.rs (hunk 3 above)
Comment (gila) [reply mayastor/src/nexus_uri.rs:103]: I'm not sure what you are suggesting..

File: mayastor/src/pool.rs (hunk 4 above)
Comment (gila) [reply mayastor/src/pool.rs:607]: Likely yes, but all these error variants, all get smooshed down into the same error anyway (internal or not found) so I'd like to address that in the follow-up releases were we reduce the error variants.

File: mayastor/src/core/bdev.rs (hunk 2 above)
Comment (gila) [reply mayastor/src/core/bdev.rs:126]: Yeah arguably, the share trait is not the right place for it and it should be just a method in the Impl block but im not sure how that will playout with replica's later

File: mayastor/src/nexus_uri.rs (hunk 3 above)
Comment (tiagolobocastro) [reply mayastor/src/nexus_uri.rs:103]: You've added the trait PartialEq for &Bdev. So now the trait PartialEq for Bdev could use make use of the &Bdev version becoming a simple bdev == uri on line 103?

Comment (gila) [reply mayastor/src/nexus_uri.rs:103]: as discussed we cant :+1:

**Comment by gila** [comment]
I've tested the following with k8s:

```yaml
 $ cat pool.yaml
apiVersion: "openebs.io/v1alpha1"
kind: MayastorPool
metadata:
  name: pool-2
  namespace: mayastor
spec:
  node: ksnode-2
  disks: ["/dev/sdb"]
---
apiVersion: "openebs.io/v1alpha1"
kind: MayastorPool
metadata:
  name: pool-3
  namespace: mayastor
spec:
  node: ksnode-3
  disks: ["aio:///dev/sdb"]
```

```bash
./bin/mayastor-client -a ksnode-2 bdev list
{
  "bdevs": [
    {
      "aliases": "",
      "blk_size": 512,
      "claimed": true,
      "claimed_by": "lvol",
      "name": "/dev/sdb",
      "num_blocks": 2097152,
      "uri": "",
      "uuid": "565e822a-6993-47f0-bb1b-c9668fff90b0"
    }
  ]
}
/bin/mayastor-client -a ksnode-3 bdev list
{
  "bdevs": [
    {
      "aliases": "aio:///dev/sdb",
      "blk_size": 512,
      "claimed": true,
      "claimed_by": "lvol",
      "name": "/dev/sdb",
      "num_blocks": 2097152,
      "uri": "aio:///dev/sdb?uuid=08ed0dc0-7028-4186-8e7a-5eb4e10b05a5",
      "uuid": "08ed0dc0-7028-4186-8e7a-5eb4e10b05a5"
    }
  ]
}
 ./bin/mayastor-client -a ksnode-3 pool list
NAME   STATE    CAPACITY USED DISKS
pool-3 online 1069547520    0 aio:///dev/sdb
```

**Pull request merged**